woocommerce/woocommerce
# Product Collection - Show product picker in Editor when collection requires a product but it doesn't exist

### Changes proposed in this Pull Request:





Part of #44877 

This PR introduces a product picker in Editor when a collection is added which requires a product context but it doesn't exist in current template/page/post etc. This way, merchant can manually provide a product context to the collection. Here is how it looks like:

![image](https://github.com/user-attachments/assets/9ea19cdb-4230-4085-be85-8664ada453ba)

To implement this, I have added `productReference` field to `query` attribute. This way, when product is manually selected by merchant, it will be stored in `productReference` attribute. This attribute will also be passed to `rest_product_query` hook to change the `WP_Query` based on selected product. Here is how it will look like:

```tsx
query: {
  productReference: number | undefined;
}
```

**What isn't covered in this PR?**
- Once product is selected, merchant should be able to change the selected product using Inspector control (More info in issue #44877). This will be covered in follow-up PR in parallel while this PR is being reviewed. Both PRs will be merged together.



### How to test the changes in this Pull Request:

**Pre-requisite:** Install this plugin - 
[register-product-collection-tester.zip](https://github.com/user-attachments/files/16362323/register-product-collection-tester.zip)

**Case 1: Adding "My Custom Collection - Product Context" to a place where product context doesn't exist**

1. Create a new post and add Product Collection block.
2. Choose "My Custom Collection - Product Context" block.
3. Verify that the product picker is shown in the Editor.
![image](https://github.com/user-attachments/assets/9ea19cdb-4230-4085-be85-8664ada453ba)

4. Search and select a product. As soon as a product is selected, verify that:
  - Product Collection should render with products.
  - The preview label is not shown when the block is selected.
  - Following _should be verified by developers only_ :
    - Verify that inside code, there is `productReference` in `query` attribute, whose value is the id of selected product i.e. `productReference: 26`
    - Verify in Network tab of REST API request that `productReference` is passed (see screenshot below). This will be accessible in `rest_product_query` hook.
![image](https://github.com/user-attachments/assets/c8a4ebbe-c1b8-4ce4-aa8d-70a9f9e191c8)

5. Follow step 1-4 for other places where product context doesn't exist i.e.
  - Page
  - Templates like Product Catalog, Products by Category, Order Confirmation etc.

**Case 2: Adding "My Custom Collection - Product Context" on Single Product template**

1. Go to Single Product template and add Product Collection block.
2. Choose "My Custom Collection - Product Context" block.
3. Verify that product picker is not shown in Editor.
4. Verify that Product Collection renders with products and preview label is shown when block is selected. On hover over the preview label, it should show a tooltip message. (See screenshot below)
![image](https://github.com/user-attachments/assets/8cb1f158-bb73-4366-8688-74c31d95a8d3)

**Case 3: Adding "My Custom Collection - Product Context" on Specific Product template**

1. Create a new Specific Product template (See screenshot below)
![image](https://github.com/user-attachments/assets/7cbfe19e-e97e-4ca7-988c-32a8ee1a3fd4)

2. Go to specific product template that you created in step 1 and add Product Collection block.
3. Choose "My Custom Collection - Product Context" block.
4. Verify that product picker is not shown in Editor.
5. Verify that Product Collection renders with products and preview label is not shown. 
  - **Note:** Preview label isn't shown because we have required product context in specific product template.

**Case 4: Collection which doesn't require product context, shouldn't show product picker**

1. Create a new post and add Product Collection block.
2. Choose any other collection except "My Custom Collection - Product Context".
3. Verify that product picker is not shown in Editor.

**Case 5: Using "Choose collection" button from Toolbar**

1. Create a new post and add Product Collection block.
2. Choose "My Custom Collection - Product Context"
3. Verify that product picker is shown in Editor.
4. Select a product and verify that collection renders with products.
5. Now click on "Choose collection" button from Toolbar and choose any other collection.
6. Verify that product picker is not shown for other collections.
7. Again choose "My Custom Collection - Product Context" and verify that product picker is shown.

**Case 6: Preview label is showing when required context is missing on Editor side but available on frontend**

As this PR also touches the code responsible for showing the preview label, we should also test this. I recommend testing "Case 3" to "Case 6" from PR #49796.





### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Product Collection - Show product picker in Editor when collection requires a product but not available

<details>
A collection can define if it requires a product context. This can be done using `usesReference` argument i.e.
```tsx
__experimentalRegisterProductCollection({
  ...,
  usesReference: ['product'],
)
```

When product context doesn't exist in current template/page/post etc. then we show product picker in Editor. This way, merchant can manually provide a product context to the collection.
</details>
<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
